doc/read-the-docs-site/adr/0006-indexer-resuming-strategy.rst

```diff
@@ -45,16 +45,89 @@ Decision
   In the current Marconi interface, it is actually just ``securityParam + 1`` as we assume that all rollbackable blocks *can* be fully stored on disk because of the disk flush.
   We assume that all the data contained in the indexer is consistent for all the point returned by an indexer.
 
-* We will have each indexer start the node-to-client chain-sync protocol at different points in time.
-  However, the ``Coordinator`` will make sure that all indexers are synced until a common point, and advance at the same speed (i.e. they can only request the next block once all the indexers have finished processing the current block).
+* We will start a single node chain-sync client for all indexers instead of one node chain sync client per indexer.
 
-* We will use the full set of points provided by ``resumableFromStorage`` of an indexer as resuming points for the node-to-client chain-sync protocol.
-  These points need to ordered in descending order so that protocol can priotise the selection of the latest ones.
+* We will implement the following resuming logic.
+
+  In order to start the chain-sync client, we need to provide to provide intersection points (or resuming points).
+  Intersection points represent the points from which we want to start syncing from the local node.
+  We will provide as intersection points the points returned from the following pseudo-code:
+
+  .. code-block:: haskell
+
+    -- | Query the resumable points of each indexer then select the resumable
+    -- points of the indexer that has the lowest immutable point. If two indexer
+    -- have the same lowest immutable point, select the one which has the lowest
+    -- overall point (immutable or not).
+    selectIntersectionPoints indexers = do
+      resumingPoints <- mapM queryResumingPoints indexers
+      head $ sortOn (\rps -> (maximum $ filter isImmutablePoint rps, maximum rps)) resumingPoints
+
+  Next, we need a function which processes each ``ChainSyncEvent`` returned by the chain-sync client for each indexer.
+  The pseudo-code is as follows:
+
+  .. code-block:: haskell
+
+    -- Preconditions:
+    --   * The 'initialLastSyncPoints' param should be sorted in ascending order on SlotNo
+    --   * Each element of the 'initialLastSyncPoints' param should satisfy the property 'getSlotNo point == getSlotNo (point + 1) + 1'.
+    applyChainSyncEvent initialLastSyncPoints previousChainSyncEventPoint event indexer = do
+      case event of
+        Resume point -> do
```

Comment:
What is a `Resume` event? Is it the initial rollback?

Reply:
That's right. Initially we didn't need it, but I think we need it right now because of the initial resume points we pass to the chain sync client.

Ex. let's say we have the following resuming points for indexers A and B
A: [1,2,3,4,5]
B: [4,5,6]

Given current resuming point selection, we will use `[1,2,3,4,5]` as the resuming points, and the chain-sync client will probably select the point 5.

So, given current implementation, the first event will be a `RollBackward`. However, we don't want B to rewind to point 5, we want to _ignore_ this RollBackward. However, how do we tell B to ignore this initial RollBackward? My solution was to introduce a Resume event.

Do you see a better way?

Reply:
Not at the moment, I don't like to add one type of event that we need to carry along during all the process while it can happen only at startup. Maybe handling the first Rollback in a different function? (`resume` would call `initiateChainSync` that would deal with the initial rollback before passing it to `applyChainSyncEvent`?)

Reply:
In the end, that type will be conformant with consensus. They also have a different type for resuming and for rollbacks. It's just that the dev that worked on cardano-streaming decided to create an event that combines the two, but the decision was arbitrary afaik.

Reply:
> Maybe handling the first Rollback in a different function? (resume would call initiateChainSync that would deal with the initial rollback before passing it to applyChainSyncEvent?)

The problem is that you don't know which point the local node will use to initiate the chain sync protocol, so that won't work.

Reply:
We still know that the first event is a rollback. Although, if Consensus already expose a specific event for resuming, we can use it as well.
In that case though, the remaining question (outside of the scope of this ADR) is probably what will we do with the `Resume` event in the general case (we may throw an error for an unexpected event or restart the resuming protocol though.
Ok, let's go for it.

Reply:
Reading back on this, your solution works as well. IIUC, you want to pass a callback function as a parameter to the `withChainSyncStreaming` function which is called on the first resume event, instead of handling a `Resume` event for every worker.

Reply:
yes, but it's implementation details, and can maybe be omitted in the adr?

Reply:
so we can go with one or the other, as long as it is correct.

Reply:
Yep let's omit it for the ADR. The principle is the same.